File: lib/haveapi/validators/acceptance.rb
module HaveAPI
  # Accepts a single configured value.
  #
  # Short form:
  #   string :param, accept: 'value'
  #
  # Full form:
  #   string :param, accept: {
  #     value: 'value',
  #     message: 'the error message'
  #   }
  class Validators::Acceptance < Validator
    name :accept
    takes :accept

    def setup
      if simple?
        @value = take

      else
        @value = take(:value)
      end

      @message = take(:message, "has to be #{@value}")
    end

    def describe
      {
          value: @value,
          message: @message,
      }
    end

    def valid?(v)
      v == @value
    end
  end
end

File: lib/haveapi/validator/acceptance.ex
defmodule HaveAPI.Validator.Acceptance do
  use HaveAPI.Validator

  def init(opts) do
    %{
      value: opts[:value],
      message: opts[:message] || "has to be '#{inspect(opts[:value])}'"
    }
  end

  def validate(%{value: v}, v, _params), do: :ok
  def validate(opts, _v, _params), do: {:error, [opts.message]}
end
